lib/Model/CommonTypeMethods.cpp

```diff
@@ -423,6 +570,62 @@ const model::Type *Common<CRTP>::skipConstAndTypedefs() const {
   }
 }
 
-// Force instantiation for the cases we care about.
-template class model::CommonTypeMethods<model::TypeDefinition>;
+// Don't forget about comparisons.
+
+template<typename CRTP>
+std::strong_ordering
+Common<CRTP>::operator<=>(const Common<CRTP> &Another) const {
+  if constexpr (std::is_same_v<CRTP, model::TypeDefinition>) {
+    // I feel like keys are descriptive enough, but the shallow comparison could
+    // be causing problems I'm not thinking of.
+    // TODO: evaluate whether we want to compare definitions deeply instead.
+    return get().key() <=> Another.get().key();
+  } else {
```

Comment:
The following pattern
```c++
if constexpr (std::is_same_v<CRTP, model::TypeDefinition>) {
  // ...
} else {
  // ...
}
```
is way too common.

Are you 100% sure that you shouldn't just reimplement each method manually in `TypeDefinition` and making a `virtual` method for each `model::Type` inheritor?

You really have to prove there's something to share here. From what I see, we're centralizing the implementations but then we're using loads of `if`s to special case them all over again.

Reply:
Yeah, it is kind of a grey area.

But if you look closer, not every overload is necessary for every helper. Sure, wide stuff like `size` has special behavior for every single one, but stuff like `tryGetAsTypedef` only cares about `typedef`s and `defined` types, silently ignoring everything else.

The point here is that I'd rather they be self-contained, as in I prefer said `tryGetAsTypedef` to be readable all at once, instead of having to jump through multiple hoops to see bits and pieces of its implementation. IMO, this makes it really simple to read: if it's a pointer do X, if it's an array do Y and so on. All neatly in one place.

> Are you 100% sure that you shouldn't just reimplement each method manually in `TypeDefinition` and making a virtual method for each `model::Type` inheritor?

Both have their own advantages and disadvantages. I prefer this: every decision that can be made at compile time is, so better error reporting when something goes wrong. As a bonus, it also guarantees that every method is implemented for every possible type - be it some "if'ed out" special behavior or fall-through.

> You really have to prove there's something to share here

Which ones have every cases provided? `size` family, partially (only types, not definitions) `skipToDefinition` family, `stripConstness` kinda (but I'm not even sure we want to keep it here in the first place) and `operator<=>`, but that guy is its own special can of worms.

All the others share at least something.

Reply:
> I don't think that's very beneficial.

I guess that's a valid difference of the view. For me, personally, the verification is kind of hard to internalize mostly because it is scattered so wildly.

With this, I tried to make it readable as possible. But I guess I'm open to option, one way or another this is just an implementation detail. The interface is what's really important.

Reply:
I reviewed the whole file, I think all the methods containing even a single `if constexpr` should be pushed to `model::TypeDefinition` and `model::Type`. I see the benefit of enforcing them to have the same exact prototype on both sides, but I don't think this extra complexity is called for. I prefer a straightforward method where I expect it to be. Other people (not having this discussion) will have to use this code, let's strive for simplicity and predictability.

Let's see what remains, but it could well mean that the whole `CommonTypeMethods` can go away.

Reply:
Done!